packages/mongodb-dialects/java-driver/src/main/kotlin/com/mongodb/jbplugin/dialects/javadriver/glossary/JavaDriverDialectParser.kt

```diff
@@ -1,33 +1,179 @@
 package com.mongodb.jbplugin.dialects.javadriver.glossary
 
-import com.intellij.psi.PsiElement
-import com.intellij.psi.PsiMethod
-import com.intellij.psi.PsiMethodCallExpression
+import com.intellij.psi.*
 import com.intellij.psi.util.PsiTreeUtil
 import com.mongodb.jbplugin.dialects.DialectParser
+import com.mongodb.jbplugin.mql.Namespace
 import com.mongodb.jbplugin.mql.Node
-import com.mongodb.jbplugin.mql.components.HasCollectionReference
+import com.mongodb.jbplugin.mql.components.*
+import com.mongodb.jbplugin.mql.toBsonType
 
 object JavaDriverDialectParser : DialectParser<PsiElement> {
-    override fun isCandidateForQuery(source: PsiElement): Boolean =
-        (source as? PsiMethodCallExpression)?.findMongoDbCollectionReference(source.project) != null
+    override fun isCandidateForQuery(source: PsiElement): Boolean {
+        if (source !is PsiMethodCallExpression) {
+// if it's not a method call, like .find(), it's not a query
+            return false
+        }
+        val sourceMethod = source.resolveMethod() ?: return false
 
-    override fun attachment(source: PsiElement): PsiElement =
-        (source as PsiMethodCallExpression).findMongoDbCollectionReference(source.project)!!
+        if ( // if the method is of MongoCollection, then we are in a query
+            sourceMethod.containingClass?.isMongoDbCollectionClass(source.project) == true
+        ) {
+            return true
+        }
+
+        if ( // if it's any driver class, check inner calls
+            sourceMethod.containingClass?.isMongoDbClass(source.project) == true ||
+            sourceMethod.containingClass?.isMongoDbCursorClass(source.project) == true
+        ) {
+            val allChildrenCandidates = PsiTreeUtil.findChildrenOfType(source, PsiMethodCallExpression::class.java)
+            return allChildrenCandidates.any { isCandidateForQuery(it) }
+        }
+
+        return false
+    }
+
+    override fun attachment(source: PsiElement): PsiElement = source.findMongoDbCollectionReference()!!
 
     override fun parse(source: PsiElement): Node<PsiElement> {
-        val owningMethod =
-            PsiTreeUtil.getParentOfType(source, PsiMethod::class.java)
-                ?: return Node(source, emptyList())
-        val namespace = NamespaceExtractor.extractNamespace(owningMethod)
-
-        return Node(
-            source,
-            listOf(
-                namespace?.let {
-                    HasCollectionReference(HasCollectionReference.Known(namespace))
-                } ?: HasCollectionReference(HasCollectionReference.Unknown),
-            ),
-        )
+        val namespace = NamespaceExtractor.extractNamespace(source)
+        val collectionReference = namespaceComponent(namespace)
+
+        val currentCall = source as PsiMethodCallExpression? ?: return Node(source, listOf(collectionReference))
```

Comment:
This specific cast is like an instance of and cast at the same time. Being a nullable type, if the casting is not successful it will return null, going through the return expression on the right side.

Reply:
ohh that's interesting - thanks for the context